lib/raven/breadcrumbs.rb

```diff
@@ -0,0 +1,81 @@
+module Raven
+  class Breadcrumb
+    attr_accessor :category, :data, :message, :level, :timestamp, :type
+
+    def initialize(init = {})
+      @category = nil
+      @data = {}
+      @level = nil
+      @message = nil
+      @timestamp = Time.now.to_i
+      @type = nil
+
+      yield self if block_given?
+    end
+
+    def to_hash
+      {
+        :category => @category,
+        :data => @data,
+        :level => @level,
+        :message => @message,
+        :timestamp => @timestamp,
+        :type => @type
+      }
+    end
+  end
+end
+
+module Raven
+  class BreadcrumbBuffer
+    def self.current
+      Thread.current[:sentry_breadcrumbs] ||= new
+    end
+
+    def self.clear!
+      current
+    end
+
+    def initialize(size = 100)
+      @count = 0
+      @pos = 0
+      @size = size
+      @buffer = Array.new(size)
+    end
+
+    def record(crumb = nil)
+      if block_given?
+        if crumb.nil?
+          crumb = Breadcrumb.new
+        end
+        yield(crumb)
+      end
+      @buffer[@pos] = crumb
+      @pos = (@pos + 1) % @size
+      @count += 1
+    end
+
+    def peek
+      return @buffer[@pos - 1]
+    end
+
+    def each
```

Comment:
Ideally this would function like blocks, but there might be some concerns there. Alternatively a better name is good enough (not sure whats standard).